common/template_string.h

```diff
@@ -0,0 +1,79 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_COMMON_TEMPLATE_STRING_H_
+#define CARBON_COMMON_TEMPLATE_STRING_H_
+
+#include "llvm/ADT/StringRef.h"
+
+namespace Carbon {
+
+// Represent a compile-time string in a form suitable for non-type template
+// arguments.
+//
+// These arguments are required to be "structural", and so we copy the string
+// contents into a public array of `char`s.
+//
+// Designed to support implicitly deduced construction from a string literal
+// template argument. This type will implicitly convert to an `llvm::StringRef`
+// for accessing the string contents, and also provides a dedicated `c_str()`
+// method to access the string as a C-string.
+//
+// Example usage:
+// ```cpp
+// template <TemplateString Str> auto F() -> void {
+//   llvm::cout() << Str;
+// }
+//
+// auto Example() -> void {
+//   F<"string contents here">();
+// }
+// ```
+template <int N>
+struct TemplateString {
+  // Constructs the template string from a string literal.
+  //
+  // Intentionally implicit conversion from string literals for use as a
+  // non-type template parameter.
+  //
+  // The closest we can get to explicitly accepting a string literal is to
+  // accept an array of `const char`s, so we additionally use Clang's constexpr
+  // `enable_if` attribute to require the array to be usable as a C-string with
+  // the expected length. This checks both for nul-termination and no embedded
```

Comment:
Two `l`s seems to be the more common https://en.wikipedia.org/wiki/Null-terminated_string
```suggestion
  // the expected length. This checks both for null-termination and no embedded
```